### PR TITLE
🪟 Add catalog changes modal on schema refresh

### LINKED ISSUE #13329: Present changes in catalog to the user
When a user uses the "Refresh catalogue" functionality on an existing saved connection, we want to show a modal containing the changes in the catalog that the user need to confirm.

The backend API for this is tracked in #13232 

**Figma mocks**:https://www.figma.com/proto/liLQhcbpVHiEDW18kiXQe3/01_03_CONNECTIONS?scaling=min-zoom&page-id=1%3A3&node-id=198%3A3822&starting-point-node-id=112%3A320
^ updated figma link 6/22

The behavior should be as follows:

* The user presses the refresh button, which would do a request towards the `web_backend/connections/get` endpoint, which will return the diff format outlined in the BE issue
* If there is a diff available, the FE will show a modal with just a confirm dialog as shown in the designs. The confirm button will only hide the modal. There is no further BE calls happening.
* The modal currently showing the saving the connection should no longer be shown.
* Should still send the full catalog to the backend

### PR DESCRIPTION
**Pull request by teallarson**
## What
closes https://github.com/airbytehq/airbyte/issues/13329

Adds a catalog change modal on source schema refresh on the connection replication page.

## How
The `web_backend/connection/read` api now sends back a catalog diff (optionally).  If there are transforms listed in the diff, the modal appears.

## Recommended reading order
I would go highest-level to most-specific:
1. `ReplicationView.tsx` shows where it is implemented
2. `CatalogDiffModal.*`
3. `CatalogDiffSection.*`
4. `CatalogDiffRow.*`
5. `CatalogDiffAccordion.*`
6. `FieldTable.*`
7. `FieldRow.*`
8. others

## Testing
1. Set up a source that you have access to make changes to (ie: will be able to change its schema, such as a local Postgres db you can add/remove tables and columns in)
2. Add it to a connection.
3. Change the source’s schema.  These are the changes detected by this catalog diff:
  i. Streams added 
  ii. Streams removed
  iii. Streams “changed”, which means a field was (a) added (b) removed or (c) changed types

### CODE REVIEW

File: airbyte-webapp/src/components/ImageBlock/ImageBlock.tsx
Comment (tealjulia) [inline airbyte-webapp/src/components/ImageBlock/ImageBlock.tsx:32]: Is `aria-label` right here?  I believe that's what you said, though I can't remember why that and not `aria-description`.  This isn't an interactive element.

Good news in general: We can now start adding `aria-labels` other places that use this component as well (this is used to render logos on our list pages, the little circles with the numbers in it on list pages, etc.)

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx:22]: I have some concerns about using a table vs flex rows and columns. Tables tend to be more difficult to get the width to match consistently, and this is not necessarily rendering as a table, either.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx:30]: should this be a forEach?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx:58]: If it's a loading button, probably needs a loading prop set.

**Review by edmundito — CHANGES_REQUESTED** [review]
Did a pass on the code but have not tested yet. Things are looking good so far, but pointed out some nits and some things that would make things more maintainable.

One thing that I didn't mention in all css files is that instead of hard coding margins and padding it should use the spacing variables.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.module.scss
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.module.scss:4]: It's unique for this modal built on top of the Headless UI "disclosure" component.  I didn't feel like there was enough I did on top of what Headless UI already provided to implement a generic Airbyte "Accordion", although that was my original plan.  Happy to revisit the idea if it seems worthwhile.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx:22]: My thought around doing it as a table was accessibility.  Can get similar functionality with flex rows/columns and aria-labels if that seems better?

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:171]: If we'll wait for https://github.com/airbytehq/airbyte/pull/14634 to be merged, we can use the ModalService introduced in this, to show this modal more via the new imperative API once the Source Scheme refresh call is done, which could improve this (since in general no state management needed anymore for it).

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.tsx
Comment (timroes) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.tsx:3]: :rocket:

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (tealjulia) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:171]: ModalService sounds great.

File: airbyte-webapp/src/components/Modal/ModalBody.module.scss
Comment (edmundito) [inline airbyte-webapp/src/components/Modal/ModalBody.module.scss:8]: `padding: 0` also works.

File: airbyte-webapp/src/components/Modal/ModalBody.tsx
Comment (edmundito) [inline airbyte-webapp/src/components/Modal/ModalBody.tsx:7]: A better name for the boolean be `padded` since `padding` may imply that I could pass the CSS style

File: airbyte-webapp/src/components/Modal/ModalBody.tsx
Comment (edmundito) [inline airbyte-webapp/src/components/Modal/ModalBody.tsx:12]: alternatively you can set it to default to `true` in the props destructuring

File: airbyte-webapp/src/hooks/services/Modal/ModalService.tsx
Comment (edmundito) [inline airbyte-webapp/src/hooks/services/Modal/ModalService.tsx:49]: The modal could still cancel here. Is this intended?

File: airbyte-webapp/src/locales/en.json
Comment (edmundito) [reply airbyte-webapp/src/locales/en.json:340]: TODO in this PR?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffAccordion.tsx:41]: These divs could be extracted together into their own component so that this Accordion component remains pretty small.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx
Comment (edmundito) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx:34]: It's fine to leave it with the table implementation until it becomes a problem.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx:57]: When applying the `syncMode`, it checks if the `streamDescriptor` is defined, but here in the key it is now. Is the `streamDescriptor` always defined?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/FieldRow.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/FieldRow.tsx:56]: If there is no oldType or newType, it will not render a table cell for this row. What will the table look like? Should it render an empty table cell instead?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/FieldSection.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/FieldSection.tsx:32]: Should only render `<ul>` if `streams.length > 0`

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/StreamRow.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/StreamRow.tsx:47]: Similar to question to earlier, if the rows mismatch how would the table render? It may be that the table cells should always render and instead check if the content is available?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/utils/utils.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/utils/utils.tsx:1]: this file can just be `CatalogDiffModal/utils.ts`

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx:20]: Could this whole modal be easily tested in a storybook with mock data?

**Review by edmundito — COMMENTED** [review]
Still looking great. I spotted a few questions here and there. I have not checked it out and tested it yet.

File: airbyte-webapp/src/hooks/services/Modal/ModalService.tsx
Comment (tealjulia) [reply airbyte-webapp/src/hooks/services/Modal/ModalService.tsx:49]: In general, our modals remain unchanged in that we should be able to cancel them by clicking "cancel" or pressing escape. 

Setting `onClose` as undefined here when the `preventCancel` property is true should prevent the modal from closing via cancel.  That's what I saw when testing it manually (both by clicking outside the modal and hitting "escape" on the keyboard).  Is there something I am not thinking of there?

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffSection.tsx:57]: it is!  cleaned up.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx:20]: Yes, but it will require [mocking the context](https://storybook.js.org/docs/react/writing-stories/decorators#context-for-mocking) as this component is only ever rendered using the modal service.

I think it's worth adding and starting to add some of our services/contexts to storybook to expand our library, but it feels like a separate PR given the scope of the work.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffFieldTable.module.scss
Comment (timroes) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffFieldTable.module.scss:28]: Since there is no classname part of that and all selectors are matching on elements only, this should apply to all tabled on the whole page (once the modal has been loaded). I think we should make sure to nest this under a class that we actually apply in the component (which likely would also get rid of the eslint warnings you were seeing and ignored in the component).

File: airbyte-webapp/src/hooks/services/Modal/ModalService.tsx
Comment (timroes) [reply airbyte-webapp/src/hooks/services/Modal/ModalService.tsx:49]: I think if the implementation calls `onCancel` we would still allow to cancel the dialog, and only prevent the "self closing feature" of it. Though there might be a better property name than `preventCancel` to express that maybe, but I don't have any better ideas than `preventCancel`.

**Comment by dizel852** [comment]
App crashes after clicking on the "Refresh source schema" button

https://www.loom.com/share/e943905d4d5f4cb5911786efdd6c735a

**Comment by tealjulia** [comment]
> App crashes after clicking on the "Refresh source schema" button
> 
> https://www.loom.com/share/e943905d4d5f4cb5911786efdd6c735a

Ah!  Addressed in the latest commit.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/CatalogDiffModal.tsx:20]: Spent some time with this Wednesday -- not easily, but I can continue working on it (maybe in a separate pr?)

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffFieldTable.module.scss
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/DiffFieldTable.module.scss]: ✅ nested

**Review by dizel852 — APPROVED** [review]
Overall - code LGTM.
Also tested locally once again - all work as should 👍🏻

**Comment by tealjulia** [comment]
@edmundito when you get a chance, it looks like this is blocked until you approve that I made the requested changes.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/utils.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/CatalogDiffModal/utils.tsx:11]: This does not need a reduce. The forEach should be ok since it's already mutating the sortedDiff object.

**Review by edmundito — COMMENTED** [review]
Code is looking really good. I checked out the branched and had a call with @tealjulia to show some alignment issues in the modal. I also noticed that the branch was not up to date with recent changes and may be worth syncing that up before merge.

File: airbyte-webapp/src/views/Connection/CatalogDiffModal/components/StreamRow.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/CatalogDiffModal/components/StreamRow.tsx:50]: Moved the cell out of that logic so it renders an empty cell if they're not passed in.

**Comment by tealjulia** [comment]
changes addressed and branch rebased.  
todo: address horizontal scrolling due to iconblock.

**Review by edmundito — APPROVED** [review]
Tested locally and changes look good. Just had a nit, and looks like there's some style that needs to be cleaned up that is causing a lint error.

**Pull request merged**
